### PR TITLE
CLOUDSTACK-8855 Improve Error Message for Host Alert State and reconnect host API.

### PR DESCRIPTION
**Pull request by DaanHoogland**
this closes #837

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@rafaelweingartner I can't add you as a reviewer. Are you member of the apache organisation?

**Comment by rhtyd** [comment]
@DaanHoogland can you fix the build failures?

**Comment by rafaelweingartner** [comment]
@DaanHoogland that is probably because I am not in the Gitbox experiment.
It is great that you picked this up, I was thinking on doing that myself, but with my current workload I would not have time to do this now.

I would say that here I see the same problem as in #837. It feels so simple to create a new column in "AlertVO" table to add the content. I am not going to repeat the discussion we had there here. So, as long as others are ok with it.

With the current state of the PR I give a -0.

**Comment by DaanHoogland** [comment]
@rafaelweingartner @marcaurele @rhtyd I am not picking this up. I just 're-opened' this because there is no movement. Please send me your PR if you want to see anything changed.

**Comment by rafaelweingartner** [comment]
@DaanHoogland ok...
I will try to squeeze this in next week.

**Comment by rafaelweingartner** [comment]
@DaanHoogland I tried to push these changes to your branch, but it seems that I do not have permissions to do so. Shall I open a PR for your branch?

**Comment by DaanHoogland** [comment]
Hey @rafaelweingartner If I new how i would a few seconds ago have given you access, you know that ;)
please send PRs so I am forced to look at them.

**Comment by rafaelweingartner** [comment]
@DaanHoogland PR created.
@marcaurele do you have some other code change suggestion to apply here?

File: engine/orchestration/src/com/cloud/agent/manager/AgentManagerImpl.java [hunk 1]
```diff
@@ -994,33 +994,28 @@ public Answer easySend(final Long hostId, final Command cmd) {
...
         if (host == null || host.getRemoved() != null) {
-            s_logger.warn("Unable to find host " + hostId);
-            return false;
+            throw new CloudRuntimeException("Unable to find host " + hostId);
         }
 
         if (host.getStatus() == Status.Disconnected) {
-            s_logger.info("Host is already disconnected, no work to be done");
-            return true;
+            throw new CloudRuntimeException("Host is already disconnected, no work to be done");
```
Comment (marcaurele) [inline engine/orchestration/src/com/cloud/agent/manager/AgentManagerImpl.java]: This looks wrong to me, why throwing an exception when nothing has to be done?

File: plugins/network-elements/netscaler/src/com/cloud/network/element/NetscalerElement.java [hunk 2]
```diff
@@ -759,7 +759,11 @@ public void doInTransactionWithoutResult(TransactionStatus status) {
         });
         HostVO host = _hostDao.findById(lbDeviceVo.getHostId());
 
-        _agentMgr.reconnect(host.getId());
+        try {
+            _agentMgr.reconnect(host.getId());
+        } catch (Exception e ) {
+            s_logger.info("failed to reconnect host "+host);
```
Comment (marcaurele) [inline plugins/network-elements/netscaler/src/com/cloud/network/element/NetscalerElement.java]: `warn`

File: engine/orchestration/src/com/cloud/agent/manager/AgentManagerImpl.java (hunk 1 above)
Comment (rafaelweingartner) [reply engine/orchestration/src/com/cloud/agent/manager/AgentManagerImpl.java]: I agree on this one. And besides that, if we take a closer look, the guy that proposed this change, was replacing the return of boolean values with exceptions to indicate some error. It makes sense that a false return would generate an exception while trying to reconnect as the case at line 1002. This one was returning true, so it feels that we should not throw an exception here.

**Comment by rafaelweingartner** [comment]
@marcaurele thanks for your feedback here. I already created a PR to Daan's branch.

File: engine/orchestration/src/com/cloud/agent/manager/AgentManagerImpl.java [hunk 3]
```diff
@@ -994,33 +980,29 @@ public Answer easySend(final Long hostId, final Command cmd) {
...
-            s_logger.warn("Unable to find host " + hostId);
-            return false;
+        if (host.getRemoved() != null) {
+            throw new CloudRuntimeException("Host has already been removed: " + hostId);
         }
 
         if (host.getStatus() == Status.Disconnected) {
-            s_logger.info("Host is already disconnected, no work to be done");
-            return true;
+            s_logger.debug("Host is already disconnected, no work to be done: " + hostId);
```
Comment (rafaelweingartner) [inline engine/orchestration/src/com/cloud/agent/manager/AgentManagerImpl.java]: @DaanHoogland I made a mistake here. Can you fix it for me?
You need to add a `return;`, right after line 994

File: engine/components-api/src/com/cloud/agent/AgentManager.java [hunk 4]
```diff
@@ -143,7 +143,7 @@
 
     public void pullAgentOutMaintenance(long hostId);
 
-    boolean reconnect(long hostId);
+    void reconnect(long hostId) throws AgentUnavailableException;
```
Comment (DaanHoogland) [inline engine/components-api/src/com/cloud/agent/AgentManager.java]: this change is breaking things. We need to look at it. @rafaelweingartner CAManagerImpl.java needs the boolean. should we take this to 4.12? I'm not taking the time now.

Comment (rafaelweingartner) [reply engine/components-api/src/com/cloud/agent/AgentManager.java]: Hmm.. Let me try to look into it today...

Comment (rafaelweingartner) [reply engine/components-api/src/com/cloud/agent/AgentManager.java]: Is the problem during compilation or execution?

Comment (rafaelweingartner) [reply engine/components-api/src/com/cloud/agent/AgentManager.java]: Never mind. Found the problem. I will work on it.

Comment (rafaelweingartner) [reply engine/components-api/src/com/cloud/agent/AgentManager.java]: I fixed the problem and opened a PR to send this whole contribution. This makes things easier for me.
PR: #2387

**Pull request closed without merging**